gaepypi/storage.py

```diff
@@ -138,22 +141,48 @@ def split_path(self, path):
         return dict(zip(components, segments))
 
     def ls(self, path, dir_only=False):
-        padded = path if path[-1] == '/' else path+'/'
-        return [f.filename for f in gcs.listbucket(padded, delimiter='/') if f.is_dir or not dir_only]
+        padded = path if path[-1] == '/' else path + '/'
+        _, bucket_path = to_bucket_and_path(padded)
+        blobs = storage_client.list_blobs(self.bucket, prefix=bucket_path, delimiter='/')
+
+        ret = []
+        for blob in blobs:
+            ret.append(self._legacy_path(blob.name))
+
+        if dir_only:
+            ret = []
+            for p in blobs.prefixes:
+                ret.append(self._legacy_path(p))
+        return ret
```

Comment:
padded = path if path[-1] == '/' else f'{path}/'
        _, bucket_path = to_bucket_and_path(padded)
        blobs = storage_client.list_blobs(self.bucket, prefix=bucket_path, delimiter='/')

        ret = [self._legacy_path(blob.name) for blob in blobs]
        if dir_only:
            ret = [self._legacy_path(p) for p in blobs.prefixes]
        return ret

Reply:
also checkout https://pypi.org/project/pathy/, [example](https://github.com/CyCognito/mayhem/blob/0a2899aae563696b563b2a0d393b9d41f641fd91/mayhem/gcloud_storage.py#L25-L28) - you might like it :)